rbd: add support for listing watchers on an RBD image

nixpanic opened the pull request:
Checking if there are other clients interested in changes to an RBD
image is important for certain actions. For example, the Ceph-CSI
project would not want to remove an image in case it is still mounted
somewhere.

Listing the details of the watchers is not important at the moment.
Adding a watcher is only implemented for testing, and is not functional
enough to get notifications through a callback. This is something that
needs more investigation and can be added at a later point.

## Checklist
- [x] Added tests for features and functional changes
- [x] Public functions and types are documented
- [x] Standard formatting is applied to Go code

nixpanic commented:
Right, testing fails on Luminous because librbd does not have `rbd_watchers_list` yet...

nixpanic commented:
This PR is rebased on #139 , so it should not get merged before that.

@phlogistonjohn could you check the `Image.HasWatchers()` function? It does not have a matching librbd function, but just counts the number of watchers that `rbd_watchers_list()` could return. Not sure if I should return a list of strings instead, or a bool, or rename the function to `CountWatchers()`.

phlogistonjohn commented:
> This PR is rebased on #139 , so it should not get merged before that.
> 

Sure. I posted a review on that.

> @phlogistonjohn could you check the `Image.HasWatchers()` function? It does not have a matching librbd function, but just counts the number of watchers that `rbd_watchers_list()` could return. Not sure if I should return a list of strings instead, or a bool, or rename the function to `CountWatchers()`.

I generally like to stick as close to the underlying apis as possible so I think we should probably define a new type, that is functionally equivalent of rbd_image_watcher_t and return an array of those from a ListWatchers() function. If  the caller only cares about the number or lack of other watchers len() of the return will suffice. If you think that's too heavyweight I'd be ok with a CountWatchers() returning an int (which int type though?).  I don't see much value in a bool-returning function or a function that returns an array of strings and loses the other struct members from rbd_image_watcher_t.

nixpanic commented:
Replaced `HasWatchers()` with `ListWatchers()` and added Image.CleanupWatchers() so that the allocated resources can be free'd again. It required some indirection for keeping a reference to the rbd_image_watch_t array, as that type is not available in older versions of Ceph.

phlogistonjohn commented:
I don't like the extra indirection/caching and cleanup approach.
Wouldn't it be simpler to define a new go type that has the fields from rbd_image_watch_t and then convert between the two? Rough example:
```go
type ImageWatcher struct {
	Addr string
        Id int64
        Cookie uint64
}

func (image *Image) ListWatchers() ([]ImageWatcher, error) {
	// ....

	watchers := make([]C.rbd_image_watcher_t, count)
	ret = C.rbd_watchers_list(image.image, &watchers[0], &count)
	if ret != 0 {
		return nil, GetError(ret)
	}
	
        iws := make([]ImageWatcher, count)
	for i, watcher := range watchers {
		iws[i].Addr = C.GoString(watcher.addr)
                iws[i].Id = int64(watcher.id)
                iws[i].Cookie = uint64(watcher.cookie)
	}
        C.rbd_watchers_list_cleanup(&watchers[0], count)
        // ....
        return iws, nil

}
```

There are no functions that take an rbd_image_watcher_t as "input" so I don't see value in keeping the original rbd_image_watcher_t values around. Only the contents. This strikes me as a lot simpler and potentially less error prone.

nixpanic commented:
That approach is indeed pretty clean, thanks for the idea!

Just one more thing, should there be a fake `ListWatchers()` for versions of Ceph that do not support it? This version will then always just return an error. It will make adopting the feature a little easier, I think. Opinions, @phlogistonjohn @ansiwen?

phlogistonjohn commented:
> That approach is indeed pretty clean, thanks for the idea!
> 
Glad it helped.

> Just one more thing, should there be a fake `ListWatchers()` for versions of Ceph that do not support it? This version will then always just return an error. It will make adopting the feature a little easier, I think. Opinions, @phlogistonjohn @ansiwen?

I'm a bit torn personally. I like the idea of failing early when possible and there's not much earlier than failing to compile. If you leave out the rbd/watchers_mimic*.go files but leave the build flags as-is I think that go will simply fail to compile with a 'no such function' error right?

I'm sympathetic to the stub approach but worry it could mislead users or lead to more subtle errors than a compile failure.

I'd also love to hear @ansiwen's opinion on this one.

nixpanic commented:
At the moment I am also adding support for new RbdImageFeatures, as newer Ceph versions add new constants. These constants will not be available for build of go-ceph with older librbd.h versions. So, I guess this is an indication that there do not need to be backwards compatible error-not-supported functions either.

Waiting for @ansiwen his opinion on the topic before proceeding.

nixpanic commented:
Btw, Ceph-CSI supports different versions of Ceph at the moment. It also checks for watchers of an image before deletion (and possibly other actions). If there is no backwards compatible function that returns `ErrNotSupported`, it will not be possible for Ceph-CSI to support Ceph versions older than Nautilus.

So, in the end, I think providing backwards compatible functions would be helpful. Not sure yet how feature bits can be provided. These could be copied from the headers, or set to 0, or ... But that comes later.

Madhu-1 commented:
ceph-csi supports Mimic+ for RBD and Nautilus+ for cephfs more info here   https://github.com/ceph/ceph-csi#ceph-csi-features-and-available-versions

nixpanic commented:
Thanks @Madhu-1! I probably mixed up projects that have tests in Travic CI for a whole range of Ceph versions :factory:

nixpanic commented:
Dropped the backwards compatible function again. It also seems that the support for watchers was added to librbd in Ceph Mimic (thanks @phlogistonjohn for spotting that).

If projects need support for older Ceph versions, they can use build-constraints like go-ceph does. go-ceph already expects a build-tag of the Ceph version for older releases (`luminous` or `mimic`).

This PR should now be ready for serious review :tada:

phlogistonjohn commented on rbd/rbd.go:
```diff
@@ -75,6 +75,7 @@ var (
 	ErrSnapshotNoName = errors.New("RBD snapshot does not have a name")
 	ErrImageNotOpen   = errors.New("RBD image not open")
 	ErrNotFound       = errors.New("RBD image not found")
+	ErrNotSupported   = errors.New("function is not supported, Ceph version might be too old")
```
Let's drop this because it has no uses in the lib now.

phlogistonjohn commented on rbd/watchers_mimic.go:
```diff
@@ -0,0 +1,98 @@
+// +build !luminous,!mimic
```
I think you want only `!luminous` here.

nixpanic replied:
Hmm, I must have pushed something before making my last changes... This is already gone in my local branch :-/

phlogistonjohn commented on rbd/watchers_mimic.go:
```diff
@@ -0,0 +1,98 @@
+// +build !luminous,!mimic
+//
+// Ceph Nautilus is the first version that supports watchers through librbd.
```
"mimic" not "nautilus"

nixpanic replied:
This too!

phlogistonjohn commented on rbd/watchers_mimic.go:
```diff
@@ -0,0 +1,98 @@
+
+// Handle for a watcher
+// This type was added so that Image.HasWatchers() can be tested.
+type rbdUpdateHandle struct {
```
If `rbdUpdateHandle` is private and has exactly one member does it add any value over just internally using some C.uint64_t values directly? If it does, the comment doesn't really explain why.

nixpanic replied:
Ah, yes, there were some other attributes while playing with callbacks. The handle is expected to be returned to the consumer of the api, so that it can be used to remove/update the watcher later on. But as all of that is private at the moment, there is no real benefit of the type (anymore).

I'll make `rbdUpdateHandle` a plain `C.uint64_t` so that it can be used in the tests (no CGO allowed).

phlogistonjohn commented on rbd/watchers_mimic.go:
```diff
@@ -0,0 +1,98 @@
+
+	ruh := &rbdUpdateHandle{}
+
+	// TODO: passing callbacks from Go to C is tricky, see
```
It would be nice to make use of the callback infrastructure that #177 provides. :-)
I'm happy to put whatever stuff we need in for private interfaces but it would be even better to have a more complete implementation for consumers. Let me know if you want to do a chat / call to dive deeper into that.

nixpanic replied:
Yes, I want to spend some more time on reviewing #177 and adapting the pattern here. So for now it is all private, until a satisfying callback 'standard' is available :+1:

phlogistonjohn commented on rbd/watchers_mimic.go:
```diff
@@ -0,0 +1,98 @@
+// and an error are returned.
+//
+// Note:
+//   Only supported in Ceph Nautilus and newer.
```
"Nautilus" -> "Mimic" again. :-)

nixpanic replied:
Yeah, missed this one.

phlogistonjohn commented on rbd/watchers_mimic.go:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	count := C.ulong(0)
+	ret := C.rbd_watchers_list(image.image, nil, &count)
```
It just occurred to me that there's may be a chance of "race condition" between the call that fetches the size/count of watchers vs. actually fetching the list. New watchers might be added between the calls and then then the 2nd call will fail (that's OK but suboptimal IMO). There's a couple of things we might want to do do mitigate that, but I won't block this PR over this. However, if you choose not to take the task on in this PR please let's file a new issue to track it.

nixpanic replied:
yes, these kind of races are present in other parts of the code as well. There is not much to do about it, as the watchers can change after this function returns and before the application acts on it. I'd prefer to see if we can use a common way of doing the ERANGE check and replace the occurrences all together.

Reported as #185

phlogistonjohn replied:
OK, some of the function calls vary in how they work with ERANGE so it may not be easy to share code, but we can certainly think about how to approach it in a common enough way.

phlogistonjohn commented on rbd/watchers_mimic_test.go:
```diff
@@ -0,0 +1,160 @@
+// +build !luminous,!mimic
```
Drop "!mimic" again

nixpanic replied:
Thats done.

phlogistonjohn commented on rbd/watchers_mimic_test.go:
```diff
@@ -0,0 +1,160 @@
+// +build !luminous,!mimic
+//
+// Ceph Nautilus is the first version that supports watchers through librbd.
```
"mimic"

phlogistonjohn commented on rbd/watchers_mimic_test.go:
```diff
@@ -0,0 +1,160 @@
+)
+
+func TestListWatchers(t *testing.T) {
+	conn, _ := rados.NewConn()
```
I added a helper function to cut down on this test setup boilerplate (and it checks more error conditions). Your original patches probably predate that getting merged, but I'd prefer it get used to make the tests more uniform.

phlogistonjohn replied:
I should be a tad bit more helpful and say the function name I'm referring to is `radosConnect`. :-)

nixpanic replied:
Ah, yes, I'll use that!

phlogistonjohn commented on rbd/watchers_mimic_test.go:
```diff
@@ -0,0 +1,160 @@
+	assert.NoError(t, err)
+
+	name := GetUUID()
+	image, err := Create(ioctx, name, 1<<22, 22)
```
Deprecated Create function. Ideally, we switch over to CreateImage. Similarly, we want OpenImage instead of image.Open unless there's a good reason to use the deprecated APIs in this test.

nixpanic replied:
And *now* I see this comment. Yes, correcting this too :smile:

phlogistonjohn commented on rbd/watchers_mimic_test.go:
```diff
@@ -0,0 +1,160 @@
+	_, err = image.ListWatchers()
+	assert.Equal(t, ErrImageNotOpen, err)
+
+	err = image.Open(true) // readonly
```
Could I pester you to isolate sections of the test that are (mostly) independent into t.Run(...) calls? It really highlights which conditions are being tested, isolates vars, and improves error reporting on test failures (IMO). I wont hold the PR over it but it would be nice if you'd consider adding that.

nixpanic replied:
Sure, I can do that. While looking at this, I notice that it also uses the old style of create/opening images. Will correct that too.

phlogistonjohn commented on rbd/watchers_mimic_test.go:
```diff
@@ -0,0 +1,160 @@
+
+func TestUpdateUnwatch(t *testing.T) {
+	// Image.updateUnwatch() is not stable
+	t.SkipNow()
```
This test is being skipped without even a user facing message. Should we just decline adding it at all for now? Also, what's unstable about it? The API, librados crash, something else?

nixpanic replied:
No crashes, it just does not always/immediately remove the watcher. I am not sure yet how the API is supposed to work. Will spend some time during the day to test a little more.

nixpanic replied:
Actually, I'll drop most of the test parts and private clutter. Opening an image writable adds a watcher automatically, closing the image removes it. Lets keep it simple for now, and add more functionality when there is a good way for callbacks.

phlogistonjohn replied:
I didn't know that, that's even better!

nixpanic commented:
This is now cleaned up, and only contains the most minimal functionality. Please review again, thanks!

phlogistonjohn commented on rbd/watchers_mimic_test.go:
```diff
@@ -0,0 +1,116 @@
+	})
+
+	// open image read-only, as OpenImage() automatically adds a watcher
+	image, err := OpenImageReadOnly(ioctx, name, NoSnapshot)
```
I think it makes more sense to move this into the imageNotOpen func call. Just remember to close it w/in that function too.

nixpanic replied:
imageNotOpen already opens and closes the image... not sure how you mean this.

phlogistonjohn commented on rbd/watchers_mimic_test.go:
```diff
@@ -0,0 +1,116 @@
+		assert.Equal(t, 0, len(watchers))
+	})
+
+	var image2, image3 *Image
```
ditto.

nixpanic replied:
Sorry, I don't understand how you mean this either.

nixpanic commented:
Just updated the tests a little (setup with `defer`). Also removed `rbd/watchers.go` and merged its contents in `rbd/watchers_mimic.go` as it only contained the ImageWatcher type. Keeping the `_mimic` extension to show when support was introduced.

phlogistonjohn commented on rbd/watchers_mimic_test.go:
```diff
@@ -0,0 +1,115 @@
+	err = CreateImage(ioctx, name, 1<<22, options)
+	require.NoError(t, err)
+
+	var image, image2, image3 *Image
```
What I have an issue with here is that allowing the objects to be referenced outside the sub-tests means that the sub tests are dependent on each other. They should only depend on the state established at the outer function. You can ask go test to only run one sub-test and that shouldn't break.
I'm not saying image shouldn't be defined outside the sub tests just that if defined outside the subtests all mutation of the state should be done outside the subtests.  I'm willing to do a peer-to-peer session if you want to if I'm not being clear in these reviews.

nixpanic replied:
Right, at the moment the tests depend on eachother. I'll structure them a little different and make them self-contained.

nixpanic commented:
@phlogistonjohn , I think your comments have been addressed now,

phlogistonjohn commented:
Looks like a bunch of the tests are failing. I think the issue are lines like:
`defer assert.NoError(t, image.Close())`
should be:
`defer func() { assert.NoError(t, image.Close()) }()`

Only the "outer" function call is deferred, so image.Close actually gets called on that line in the first example. Wrapping everything in a function makes sure the entire "line" is deferred.

I will do a proper review once the tests are passing again.

nixpanic commented:
On Tue, Feb 25, 2020 at 07:13:08AM -0800, John Mulligan wrote:
> Looks like a bunch of the tests are failing. I think the issue are lines like:
> `defer assert.NoError(t, image.Close())`
> should be:
> `defer func() { assert.NoError(t, image.Close()) }()`
> 
> Only the "outer" function call is deferred, so image.Close actually gets called on that line in the first example. Wrapping everything in a function makes sure the entire "line" is deferred.

Right! I did wonder a little why tests were failing in CI and working
locally... Obviously I should not pass CEPH_VERSION=luminous on the
commandline :facepalm:

nixpanic commented:
This is fixed now, thanks for having a look!
